### DISCUSSION TITLE
Rule execution order

### DISCUSSION THREAD

**Discussion by picklebento** [detekt/detekt#3445] [Ideas]
In #3317, there are different arguments for running auto-correctible rules first or non-correctible rules first. 

In addition, the existing multi rule paradigm is complicating our situation.
- KtlintMultiRule is regarded as either all auto-correctible or non-autocorrectible. But in fact, it mixes auto-correctible with non-autocorrectible rules. Thus we need to split KtlintMultiRule into two sets, at least.
- KtLint rules have certain modifiers to impose orders as MaxLineLengthRule runs as the Last rule. Some long lines may not actually exceed the line length limit after mutation from other rules. I am pretty confident that this issue exists for Detekt as well.
- Parallelism is my perception of the intention of creating MultiRule, but I believe we could do much better here to figure out the strategy across the board. The argument here would be a Rule can report Multiple issues.

**Comment by BraisGabin** [comment]
Related with #2081

I still don't understand why to use ktlint inside detekt instead of ktlint directly. It works as a charm by itself. They even support baseline now!

**Reply by cortinico** [reply]
> I still don't understand why to use ktlint inside detekt instead of ktlint directly. It works as a charm by itself. They even support baseline now!

True. I believe is creating confusion and just complicating our life as maintainers. I believe is for historical reasons.

**Comment by BraisGabin** [comment]
About order between correctable and no correctable I was thinking about splitting it. If you want to correct issues you have one task. It doesn't fail because of issues and it doesn't create any report. It just fix. And then, the main task, that creates the reports and fail if there are issues etc.

I think that it will fix our current problems about what to run before. But as an user if you want to fix all the fixable and get the report or will be a bit slower because you need to run two tasks.

**Reply by cortinico** [reply]
> But as an user if you want to fix all the fixable and get the report or will be a bit slower because you need to run two tasks.

As long as it's configured that the "autoCorrect" tasks runs before the actual detekt task, it should work fine.
